File: 20180622_SDK/SDK/Sample_M50/C#_SBXPCDLL_Sample/SBXPCDLLSampleCSharp/frmTrMode.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

using System.Runtime.InteropServices;

namespace SBXPCDLLSampleCSharp
{
    public partial class frmTrMode : Form
    {
        public frmTrMode()
        {
            InitializeComponent();
        }

        const int DB_TRMODE_COUNT = 10;
        const int DB_TRMODE_BYTE_COUNT = 5;
        int[] trModeInfo;

        private void frmTrMode_Load(object sender, EventArgs e)
        {
            trModeInfo = new int[DB_TRMODE_COUNT * DB_TRMODE_BYTE_COUNT];
            InitTrMode();
            DrawTrModeList();
            dtStart.ShowUpDown = true;
            dtEnd.ShowUpDown = true;
        }

        private void InitTrMode()
        {
            for(int i = 0; i < DB_TRMODE_COUNT; i ++)
            {
                trModeInfo[i * DB_TRMODE_BYTE_COUNT] = 0;
                trModeInfo[i * DB_TRMODE_BYTE_COUNT + 1] = 0;
                trModeInfo[i * DB_TRMODE_BYTE_COUNT + 2] = 0;
                trModeInfo[i * DB_TRMODE_BYTE_COUNT + 3] = 0;
                trModeInfo[i * DB_TRMODE_BYTE_COUNT + 4] = 0;
            }
        }

        private void DrawTrModeList()
        {
            string itemString;
            lstTrMode.Items.Clear();
            for(int i = 0; i < DB_TRMODE_COUNT; i ++)
            {
                itemString = "[No.]" + String.Format("{0:D2}", i) + " ";
                itemString += "[S]";
                itemString += String.Format("{0:D2}", trModeInfo[i * DB_TRMODE_BYTE_COUNT + 1]) + ":";
                itemString += String.Format("{0:D2}", trModeInfo[i * DB_TRMODE_BYTE_COUNT + 2]) + " ";
                itemString += "[E]";
                itemString += String.Format("{0:D2}", trModeInfo[i * DB_TRMODE_BYTE_COUNT + 3]) + ":";
                itemString += String.Format("{0:D2}", trModeInfo[i * DB_TRMODE_BYTE_COUNT + 4]) + " ";
                itemString += cmbTrMode.Items[trModeInfo[i * DB_TRMODE_BYTE_COUNT]];
                lstTrMode.Items.Add(itemString);
            }
        }

        private void frmTrMode_FormClosing(object sender, FormClosingEventArgs e)
        {
            Application.OpenForms["frmMain"].Visible =true;
        }

        private void cmdExit_Click(object sender, EventArgs e)
        {
            Close();
        }

        private void lstTrMode_SelectedIndexChanged(object sender, EventArgs e)
        {
            if (lstTrMode.SelectedIndex == -1)
                return;

            int index       = lstTrMode.SelectedIndex * DB_TRMODE_BYTE_COUNT;
            int startHour   = trModeInfo[index + 1];
            int startMinute = trModeInfo[index + 2];
            int endHour     = trModeInfo[index + 3];
            int endMinute   = trModeInfo[index + 4];
            int trMode      = trModeInfo[index];

            dtStart.Value = new DateTime(2000, 1, 1, startHour, startMinute, 0);
            dtEnd.Value = new DateTime(2000, 1, 1, endHour, endMinute, 0);
            cmbTrMode.SelectedIndex = trMode;
        }

        private void cmdUpdate_Click(object sender, EventArgs e)
        {
            if (lstTrMode.SelectedIndex == -1)
                return;

            int index = lstTrMode.SelectedIndex * DB_TRMODE_BYTE_COUNT; 
            trModeInfo[index + 1]   = dtStart.Value.Hour;
            trModeInfo[index + 2]   = dtStart.Value.Minute;
            trModeInfo[index + 3]   = dtEnd.Value.Hour;
            trModeInfo[index + 4]   = dtEnd.Value.Minute;
            trModeInfo[index]       = cmbTrMode.SelectedIndex;

            DrawTrModeList();
        }

        private void cmdRead_Click(object sender, EventArgs e)
        {
            bool bRet;
            int vErrorCode = 0;

            lblMessage.Text = "Working...";
            Application.DoEvents();

            bRet = sbxpc.SBXPCDLL.EnableDevice(Program.gMachineNumber, 0); // 0 : disable

            if (!bRet)
            {
                lblMessage.Text = util.gstrNoDevice;
                return;
            }

            GCHandle gh = GCHandle.Alloc(trModeInfo, GCHandleType.Pinned);
            IntPtr AddrOftrModeInfo = gh.AddrOfPinnedObject();

            bRet = sbxpc.SBXPCDLL.GetDeviceLongInfo(Program.gMachineNumber, 5, AddrOftrModeInfo);

            if (bRet)
            {
                lblMessage.Text = "Success!";
            }
            else
            {
                sbxpc.SBXPCDLL.GetLastError(Program.gMachineNumber, out vErrorCode);
                lblMessage.Text = util.ErrorPrint(vErrorCode);
            }

            bRet = sbxpc.SBXPCDLL.EnableDevice(Program.gMachineNumber, 1); // 1 : enable

            DrawTrModeList();
        }

        private void cmdWrite_Click(object sender, EventArgs e)
        {
            bool bRet;
            int vErrorCode = 0;

            lblMessage.Text = "Working...";
            Application.DoEvents();

            bRet = sbxpc.SBXPCDLL.EnableDevice(Program.gMachineNumber, 0); // 0 : disable

            if (!bRet)
            {
                lblMessage.Text = util.gstrNoDevice;
                return;
            }

            GCHandle gh = GCHandle.Alloc(trModeInfo, GCHandleType.Pinned);
            IntPtr AddrOftrModeInfo = gh.AddrOfPinnedObject();

            bRet = sbxpc.SBXPCDLL.SetDeviceLongInfo(Program.gMachineNumber, 5, AddrOftrModeInfo);

            if (bRet)
            {
                lblMessage.Text = "Success!";
            }
            else
            {
                sbxpc.SBXPCDLL.GetLastError(Program.gMachineNumber, out vErrorCode);
                lblMessage.Text = util.ErrorPrint(vErrorCode);
            }

            bRet = sbxpc.SBXPCDLL.EnableDevice(Program.gMachineNumber, 1); // 1 : enable

            DrawTrModeList();
        }

        private void cmdLoadFile_Click(object sender, EventArgs e)
        {
            DialogResult result = OpenFileDlg.ShowDialog();
            if (result == DialogResult.Cancel) return;

            if (!util.LoadDeviceLongInfo(OpenFileDlg.FileName, trModeInfo))
            {
                MessageBox.Show("Failed to open file : " + OpenFileDlg.FileName);
                return;
            }
            DrawTrModeList();
        }

        private void cmdSaveFile_Click(object sender, EventArgs e)
        {
            DialogResult result = OpenSaveDlg.ShowDialog();
            if (result == DialogResult.Cancel) return;

            if (!util.SaveDeviceLongInfo(OpenSaveDlg.FileName, trModeInfo))
            {
                MessageBox.Show("Unexpected Error");
                return;
            }
        }
    }
}

File: 20180622_SDK/SDK/Sample_M50/VB.NET_SBXBPCSample/frmTrMode.vb
Option Strict Off
Option Explicit On
Imports System.Runtime.InteropServices

Public Class frmTrMode

    Const DB_TRMODE_COUNT As Integer = 10
    Const DB_TRMODE_BYTE_COUNT As Integer = 5
    Dim trModeInfo(DB_TRMODE_COUNT * DB_TRMODE_BYTE_COUNT - 1) As UInt32

    Private Sub frmTrMode_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        trModeInit()
        DrawTrMode()
        dtStart.ShowUpDown = True
        dtEnd.ShowUpDown = True
    End Sub

    Private Sub trModeInit()
        Dim i As Integer
        For i = 0 To DB_TRMODE_COUNT - 1
            trModeInfo(i * DB_TRMODE_BYTE_COUNT + 0) = 0
            trModeInfo(i * DB_TRMODE_BYTE_COUNT + 1) = 0
            trModeInfo(i * DB_TRMODE_BYTE_COUNT + 2) = 0
            trModeInfo(i * DB_TRMODE_BYTE_COUNT + 3) = 0
            trModeInfo(i * DB_TRMODE_BYTE_COUNT + 4) = 0
        Next
    End Sub

    Private Sub DrawTrMode()
        Dim itemString As String
        Dim i As Integer
        lstTrMode.Items.Clear()

        For i = 0 To DB_TRMODE_COUNT - 1
            itemString = "[No.]" + String.Format("{0:D2}", i) + " "
            itemString += "[S]"
            itemString += String.Format("{0:D2}", trModeInfo(i * DB_TRMODE_BYTE_COUNT + 1)) & ":"
            itemString += String.Format("{0:D2}", trModeInfo(i * DB_TRMODE_BYTE_COUNT + 2)) & " "
            itemString += "[E]"
            itemString += String.Format("{0:D2}", trModeInfo(i * DB_TRMODE_BYTE_COUNT + 3)) & ":"
            itemString += String.Format("{0:D2}", trModeInfo(i * DB_TRMODE_BYTE_COUNT + 4)) & " "
            itemString += cmbTrMode.Items(trModeInfo(i * DB_TRMODE_BYTE_COUNT))
            lstTrMode.Items.Add(itemString)
        Next
    End Sub

    Private Sub cmdRead_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmdRead.Click
        Dim bRet As Boolean
        Dim vErrorCode As Integer

        lblMessage.Text = "Working..."
        Application.DoEvents()

        bRet = sbxpc.SBXPCDLL.EnableDevice(frmMain.gMachineNumber, False)
        If Not bRet Then
            lblMessage.Text = gstrNoDevice
            Exit Sub
        End If

        Dim gh As GCHandle = GCHandle.Alloc(trModeInfo, GCHandleType.Pinned)
        Dim AddrOftimZoneInfo As IntPtr = gh.AddrOfPinnedObject()
        bRet = sbxpc.SBXPCDLL.GetDeviceLongInfo(frmMain.gMachineNumber, 5, AddrOftimZoneInfo.ToInt32())

        If bRet Then
            lblMessage.Text = "Success"
        Else
            sbxpc.SBXPCDLL.GetLastError(frmMain.gMachineNumber, vErrorCode)
            lblMessage.Text = ErrorPrint(vErrorCode)
        End If

        bRet = sbxpc.SBXPCDLL.EnableDevice(frmMain.gMachineNumber, True)

        DrawTrMode()
    End Sub

    Private Sub cmdWrite_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmdWrite.Click
        Dim bRet As Boolean
        Dim vErrorCode As Integer

        lblMessage.Text = "Working..."
        Application.DoEvents()

        bRet = sbxpc.SBXPCDLL.EnableDevice(frmMain.gMachineNumber, False)
        If Not bRet Then
            lblMessage.Text = gstrNoDevice
            Exit Sub
        End If

        Dim gh As GCHandle = GCHandle.Alloc(trModeInfo, GCHandleType.Pinned)
        Dim AddrOftimZoneInfo As IntPtr = gh.AddrOfPinnedObject()
        bRet = sbxpc.SBXPCDLL.SetDeviceLongInfo(frmMain.gMachineNumber, 5, AddrOftimZoneInfo.ToInt32())

        If bRet Then
            lblMessage.Text = "Success"
        Else
            sbxpc.SBXPCDLL.GetLastError(frmMain.gMachineNumber, vErrorCode)
            lblMessage.Text = ErrorPrint(vErrorCode)
        End If

        bRet = sbxpc.SBXPCDLL.EnableDevice(frmMain.gMachineNumber, True)
    End Sub

    Private Sub frmTrMode_FormClosing(ByVal sender As System.Object, ByVal e As System.Windows.Forms.FormClosingEventArgs) Handles MyBase.FormClosing
        frmMain.Visible = True
    End Sub

    Private Sub cmdExit_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmdExit.Click
        Close()
    End Sub

    Private Sub lstTrMode_SelectedIndexChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles lstTrMode.SelectedIndexChanged
        If lstTrMode.SelectedIndex = -1 Then
            Exit Sub
        End If

        Dim index As Integer = lstTrMode.SelectedIndex * DB_TRMODE_BYTE_COUNT
        dtStart.Value = New Date(2000, 1, 1, trModeInfo(index + 1), trModeInfo(index + 2), 0)
        dtEnd.Value = New Date(2000, 1, 1, trModeInfo(index + 3), trModeInfo(index + 4), 0)
        cmbTrMode.SelectedIndex = trModeInfo(index)
    End Sub

    Private Sub cmdUpdate_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmdUpdate.Click
        If lstTrMode.SelectedIndex = -1 Then
            Exit Sub
        End If

        Dim index As Integer = lstTrMode.SelectedIndex * DB_TRMODE_BYTE_COUNT
        trModeInfo(index) = cmbTrMode.SelectedIndex
        trModeInfo(index + 1) = dtStart.Value.Hour
        trModeInfo(index + 2) = dtStart.Value.Minute
        trModeInfo(index + 3) = dtEnd.Value.Hour
        trModeInfo(index + 4) = dtEnd.Value.Minute

        DrawTrMode()
    End Sub
End Class